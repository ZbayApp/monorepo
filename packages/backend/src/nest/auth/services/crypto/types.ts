import { Base58, SignedEnvelope } from '@localfirst/auth'

export enum EncryptionScopeType {
  ROLE = 'ROLE',
  CHANNEL = 'CHANNEL',
  USER = 'USER',
  TEAM = 'TEAM',
}

export type EncryptionScope = {
  type: EncryptionScopeType
  name?: string
}

export type EncryptedPayload = {
  contents: Base58
  scope: EncryptionScope & {
    generation: number
  }
}

export type EncryptedAndSignedPayload = {
  encrypted: EncryptedPayload
  signature: SignedEnvelope
  ts: number
  username: string
}
