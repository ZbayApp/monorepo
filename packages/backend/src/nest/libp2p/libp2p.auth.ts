import {
  ComponentLogger,
  Connection,
  PeerId,
  PeerStore,
  serviceCapabilities,
  Stream,
  Topology,
} from '@libp2p/interface'
import type { ConnectionManager, IncomingStreamData, Registrar } from '@libp2p/interface-internal'
import * as Auth from '@localfirst/auth'
import { pipe } from 'it-pipe'
import { encode, decode } from 'it-length-prefixed'

import { SigChainService } from '../auth/sigchain.service'
import { createLogger } from '../common/logger'
import { createQuietLogger } from '@quiet/logger'
import { ConnectionParams } from '3rd-party/auth/packages/auth/dist/connection/Connection'
import { Libp2pService } from './libp2p.service'
import { Libp2pEvents } from './libp2p.types'
import { abortableAsyncIterable } from '../common/utils'

export interface Libp2pAuthComponents {
  peerId: PeerId
  peerStore: PeerStore
  registrar: Registrar
  connectionManager: ConnectionManager
  logger: ComponentLogger
}

export interface Libp2pAuthStatus {
  started: boolean
  joining: boolean
}

enum JoinStatus {
  PENDING = 'PENDING',
  JOINING = 'JOINING',
  JOINED = 'JOINED',
  NOT_STARTED = 'NOT_STARTED',
}

const createLFALogger = createQuietLogger('localfirst')

export class Libp2pAuth {
  private readonly protocol: string
  private readonly components: Libp2pAuthComponents
  private sigChainService: SigChainService
  private libp2pService: Libp2pService
  private authConnections: Map<string, Auth.Connection>
  private peerConnections: Map<string, Connection>
  private bufferedConnections: { peerId: PeerId; connection: Connection }[]
  private unblockInterval: NodeJS.Timeout
  private joinStatus: JoinStatus
  private logger: ReturnType<typeof createLogger> = createLogger('libp2p:auth')
  readonly [serviceCapabilities]: string[] = ['@quiet/auth']
  readonly [Symbol.toStringTag]: string = 'lfaAuth'

  constructor(sigChainService: SigChainService, libp2pService: Libp2pService, components: Libp2pAuthComponents) {
    this.protocol = '/local-first-auth/1.0.0'
    this.components = components
    this.sigChainService = sigChainService
    this.libp2pService = libp2pService
    this.authConnections = new Map()
    this.peerConnections = new Map()
    this.bufferedConnections = []

    if (sigChainService.activeChainTeamName == null) {
      this.logger.warn('No active chain found')
      this.joinStatus = JoinStatus.NOT_STARTED
    } else {
      this.logger = this.logger.extend(sigChainService.getActiveChain().localUserContext.user.userName)
      if (sigChainService.getActiveChain()!.team == null) {
        this.joinStatus = JoinStatus.PENDING
      } else {
        this.joinStatus = JoinStatus.JOINED
      }
    }

    this.logger.info('Auth service initialized')
    this.logger.info('sigChainService', sigChainService.activeChainTeamName)

    // Set up a periodic check to process buffered connections
    this.unblockInterval = setInterval(this.unblockConnections, 5_000, this.bufferedConnections)
  }

  private emit(eventName: string, ...args: any[]) {
    this.libp2pService.emit(eventName, ...args)
  }

  // Process any connections that were buffered because we were waiting for a chain
  private async unblockConnections(conns: { peerId: PeerId; connection: Connection }[]) {
    if (this.joinStatus === JoinStatus.NOT_STARTED && this.sigChainService.activeChainTeamName != null) {
      this.logger.info(`Unblocking ${conns.length} connections now that we have an active chain`)
      this.joinStatus = JoinStatus.PENDING
    } else if (this.joinStatus !== JoinStatus.JOINED) {
      return
    }

    this.logger.info(`Unblocking ${conns.length} buffered connections now that we've joined the chain`)
    while (conns.length > 0) {
      const conn = conns.pop()
      if (conn != null) {
        await this.onPeerConnected(conn.peerId, conn.connection)
      }
    }
  }

  async start() {
    this.logger.info('Auth service starting')

    const topology: Topology = {
      onConnect: this.onPeerConnected.bind(this),
      onDisconnect: this.onPeerDisconnected.bind(this),
      notifyOnLimitedConnection: false,
    }

    const registrar = this.components.registrar
    await registrar.register(this.protocol, topology)
    await registrar.handle(this.protocol, this.onIncomingStream.bind(this), {
      runOnLimitedConnection: false,
    })
  }

  async beforeStop() {
    this.logger.info('beforeStop')
  }

  async stop() {
    this.logger.info('stop')

    // Clear the unblock interval
    clearInterval(this.unblockInterval)

    // Close all auth connections
    for (const peerId of this.authConnections.keys()) {
      this.closeAuthConnection(peerId)
    }

    this.logger.info('Libp2pAuth service stopped')
  }

  async afterStop() {
    this.logger.info('afterStop')
    if (this.sigChainService.activeChainTeamName != null) {
      await this.sigChainService.saveChain(this.sigChainService.activeChainTeamName)
    }
  }

  /**
   * Handle an incoming ephemeral stream.
   * Once the stream is processed, it is closed.
   */
  private async onIncomingStream({ stream, connection }: IncomingStreamData) {
    const peerId = connection.remotePeer
    this.logger.info(`Handling incoming ephemeral stream ${connection.id.toString()} from ${peerId.toString()}`)
    const abortController = new AbortController()

    // Process messages from the stream
    this.handleIncomingMessages(peerId, stream, abortController)
      .catch(err => {
        this.logger.error(`Error processing incoming stream from ${peerId.toString()}`, err)
        if (!abortController.signal.aborted) {
          abortController.abort(err)
        }
      })
      .finally(() => {
        stream
          .close()
          .catch(err => this.logger.error(`Error closing incoming ephemeral stream from ${peerId.toString()}`, err))
      })
  }

  /**
   * Process incoming messages by decoding the length-prefixed data and delivering
   * it to the corresponding auth connection.
   */
  private async handleIncomingMessages(peerId: PeerId, stream: Stream, abortController: AbortController) {
    await pipe(
      stream,
      source => decode(source),
      async source => {
        for await (const data of abortableAsyncIterable(source, abortController.signal)) {
          try {
            const authConn = this.authConnections.get(peerId.toString())
            if (!authConn) {
              this.logger.error(`No auth connection established for ${peerId.toString()}`)
            } else {
              authConn.deliver(data.subarray())
            }
          } catch (e) {
            this.logger.error(`Error while delivering message to ${peerId.toString()}`, e)
            if (!abortController.signal.aborted) {
              abortController.abort(e)
            }
          }
        }
      }
    )
  }

  /**
   * Send an outgoing message using an ephemeral stream.
   * This method opens a new stream, writes the encoded message, and then closes it.
   */
  private async sendMessage(peerId: PeerId, message: Uint8Array) {
    const connection = this.peerConnections.get(peerId.toString())
    if (!connection) {
      this.logger.warn(`No connection available for ephemeral stream to ${peerId.toString()}`)
      return
    }

    const abortController = new AbortController()
    try {
      this.logger.info(`Opening ephemeral outbound stream to ${peerId.toString()}`)
      const stream = await connection.newStream(this.protocol, {
        runOnLimitedConnection: false,
        negotiateFully: false,
        signal: abortController.signal,
      })
      this.logger.info(`Ephemeral stream opened to ${peerId.toString()}, sending message`)
      if (stream.status !== 'open') {
        this.logger.warn(
          `Attempted to send message to ${peerId.toString()} on ephemeral stream that had already closed`
        )
        return
      }
      await pipe([encode.single(message)], stream)
      await stream.close()
      this.logger.info(`Ephemeral stream closed to ${peerId.toString()}`)
    } catch (e) {
      this.logger.error(`Error sending ephemeral message to ${peerId.toString()}`, e)
      if (!abortController.signal.aborted) {
        abortController.abort(e)
      }
    }
  }

  /**
   * Called when a peer connects. If weâ€™re not ready to start (e.g. no active chain),
   * the connection is buffered. Otherwise we create a new auth connection and
   * store the underlying libp2p connection for ephemeral stream use.
   */
  private async onPeerConnected(peerId: PeerId, connection: Connection) {
    if (this.authConnections.has(peerId.toString())) {
      this.logger.info(`Auth connection with ${peerId.toString()} already exists`)
      return
    }
    if (this.joinStatus === JoinStatus.JOINING) {
      this.logger.warn(`Connection to ${peerId.toString()} will be buffered due to a concurrent join`)
      this.bufferedConnections.push({ peerId, connection })
      return
    }
    if (this.sigChainService.activeChainTeamName == null) {
      this.logger.warn(`No active chain found, buffering connection to ${peerId.toString()}`)
      this.bufferedConnections.push({ peerId, connection })
      return
    }

    if (this.joinStatus === JoinStatus.PENDING) {
      this.joinStatus = JoinStatus.JOINING
    }

    this.logger.info(`Peer connected (direction = ${connection.direction})! (status = ${connection.status})`)
    if (connection.status !== 'open') {
      this.logger.warn(`The connection with ${peerId.toString()} was not in an open state!`)
      return
    }

    const context = this.sigChainService.getActiveChain().context

    if (this.authConnections.has(peerId.toString())) {
      this.logger.info(
        `A connection with ${peerId.toString()} was already available, skipping connection initialization!`
      )
      return
    }

    // Create an auth connection using an ephemeral sendMessage callback.
    const authConnection = new Auth.Connection({
      context,
      sendMessage: (message: Uint8Array) => {
        // Fire-and-forget: send message using an ephemeral stream.
        this.sendMessage(peerId, message).catch(err => {
          this.logger.error(`Error in sendMessage callback for ${peerId.toString()}`, err)
        })
      },
      createLogger: createLFALogger,
    } as ConnectionParams)

    // Set up auth connection event handlers.
    authConnection.on('connected', () => {
      if (this.sigChainService.activeChainTeamName != null) {
        this.logger.debug(`Sending sync message because our chain is initialized`)
        const sigChain = this.sigChainService.getActiveChain()
        const team = sigChain.team
        const user = sigChain.localUserContext.user
        authConnection.emit('sync', { team, user })
      }
      this.emit(Libp2pEvents.AUTH_CONNECTED)
    })

    authConnection.on('disconnected', event => {
      this.logger.info(`LFA Disconnected!`, event)
      this.libp2pService.emit(Libp2pEvents.AUTH_DISCONNECTED, {
        event,
        connection,
      })
    })

    authConnection.on('joined', async payload => {
      const { team, user } = payload
      const sigChain = this.sigChainService.getActiveChain()
      this.logger.info(
        `${sigChain.localUserContext.user.userId}: Joined team ${team.teamName} (userid: ${user.userId})!`
      )
      if (sigChain.team == null) {
        this.logger.info(
          `${user.userId}: Creating SigChain for user with name ${user.userName} and team name ${team.teamName}`
        )
        sigChain.context = {
          device: (sigChain.context as Auth.InviteeContext).device,
          team,
          user,
        } as Auth.MemberContext
        sigChain.team = team
      }
      this.joinStatus = JoinStatus.JOINED
      this.unblockConnections(this.bufferedConnections)
      this.emit(Libp2pEvents.AUTH_JOINED)
      await this.sigChainService.saveChain(team.teamName)
    })

    authConnection.on('change', payload => {
      this.emit(Libp2pEvents.AUTH_STATE_CHANGED, payload)
    })

    authConnection.on('updated', async head => {
      this.logger.info('Received sync message, team graph updated', head)
      this.emit(Libp2pEvents.AUTH_UPDATED, head)
      const sigChain = this.sigChainService.getActiveChain()
      await this.sigChainService.saveChain(sigChain.team!.teamName)
    })

    // Handle errors from local or remote sources.
    authConnection.on('localError', error => {
      this.emit(Libp2pEvents.AUTH_LOCAL_ERROR, { error, connection })
    })
    authConnection.on('remoteError', error => {
      this.emit(Libp2pEvents.AUTH_REMOTE_ERROR, { error, connection })
    })

    // Store the auth connection and also the underlying libp2p connection
    this.authConnections.set(peerId.toString(), authConnection)
    this.peerConnections.set(peerId.toString(), connection)

    this.logger.info(`Auth connection established with ${peerId.toString()}`)
    authConnection.start()
  }

  private async onPeerDisconnected(peerId: PeerId) {
    if (this.authConnections.has(peerId.toString())) {
      this.logger.warn(`Auth connection with ${peerId.toString()} was disconnected`)
      this.closeAuthConnection(peerId)
    }
  }

  public closeAuthConnection(peerId: PeerId | string) {
    this.logger.info(`Attempting to close auth connection with ${peerId.toString()}`)
    const key = peerId.toString()

    // Remove the stored connection (ephemeral streams are used for each message)
    if (this.peerConnections.has(key)) {
      this.peerConnections.delete(key)
    }

    if (this.authConnections.has(key)) {
      try {
        this.authConnections.get(key)?.stop()
      } catch (e) {
        // do nothing
      }
      this.authConnections.delete(key)
    }
  }
}

export const libp2pAuth = (
  sigChainService: SigChainService,
  libp2pService: Libp2pService
): ((components: Libp2pAuthComponents) => Libp2pAuth) => {
  return (components: Libp2pAuthComponents) => new Libp2pAuth(sigChainService, libp2pService, components)
}
