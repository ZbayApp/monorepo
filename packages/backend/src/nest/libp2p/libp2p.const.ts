export const BITSWAP_PROTOCOL = '/quiet/ipfs/bitswap/1.2.0'
export const WEBSOCKET_CIPHER_SUITE = 'TLS_AES_256_GCM_SHA384'
