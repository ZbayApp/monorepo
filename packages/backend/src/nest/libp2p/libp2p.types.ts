import { PeerId, PrivateKey } from '@libp2p/interface'
import { Agent } from 'http'

export enum Libp2pEvents {
  PEER_CONNECTED = 'peerConnected',
  PEER_DISCONNECTED = 'peerDisconnected',
  NETWORK_STATS = 'networkStats',
}

export interface Libp2pNodeParams {
  peerId: CreatedLibp2pPeerId
  listenAddresses: string[]
  agent: Agent
  localAddress: string
  targetPort: number
  psk: Uint8Array
}

export type Libp2pPeerInfo = {
  dialed: string[]
  connected: string[]
}

export type Libp2pConnectedPeer = {
  address: string
  connectedAtSeconds: number
}

export type Libp2pDatastoreOptions = {
  inMemory: boolean
  datastorePath?: string
}

export interface CreatedLibp2pPeerId {
  peerId: PeerId
  privKey: PrivateKey
}
