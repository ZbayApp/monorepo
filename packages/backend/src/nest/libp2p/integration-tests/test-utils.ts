import { Test, TestingModule } from '@nestjs/testing'
import { TestModule } from '../../common/test.module'
import { getLocalLibp2pInstanceParams } from '../../common/utils'
import { Libp2pModule } from '../libp2p.module'
import { SigChainModule } from '../../auth/sigchain.service.module'
import { Libp2pEvents, Libp2pNodeParams } from '../libp2p.types'
import { createLogger } from '../../common/logger'
import { Libp2pService } from '../libp2p.service'

const logger = createLogger('libp2p:test-utils')

export const attachEventListeners = (libp2pService: Libp2pService, timeline: string[], instanceName: string) => {
  // loop over all enum Libp2pEvents and attach event listeners
  for (const event of Object.values(Libp2pEvents)) {
    if (event === Libp2pEvents.AUTH_STATE_CHANGED) {
      libp2pService.on(event, (state: any) => {
        timeline.push(`${event}:${state}`)
      })
    } else {
      libp2pService.on(event, () => {
        timeline.push(`${event}`)
      })
    }
  }
}

export const spawnTestModules = async (number: number) => {
  const modules = []
  for (let i = 0; i < number; i++) {
    const module = await Test.createTestingModule({
      imports: [TestModule, Libp2pModule, SigChainModule],
    }).compile()
    modules.push(module)
  }
  logger.info(`created ${modules.length} test modules`)
  return modules
}

export const spawnLibp2pInstances = async (
  modules: TestingModule[],
  customLibp2pInstanceParams?: Libp2pNodeParams,
  sharePsk: boolean = true
) => {
  logger.info(`creating ${modules.length} libp2p instances`)
  const singlePSK = Libp2pService.generateLibp2pPSK().fullKey
  const libp2pServices = []
  for (let i = 0; i < modules.length; i++) {
    logger.info(`creating libp2p instance ${i}`)
    const libp2pService = await modules[i].resolve(Libp2pService)
    const params = {
      ...(await getLocalLibp2pInstanceParams()),
      ...customLibp2pInstanceParams,
      instanceName: `instance${i}`,
    }
    if (sharePsk) {
      params.psk = singlePSK
    }
    logger.info(`creating libp2p instance with params:`, params)
    await libp2pService.createInstance(params)
    libp2pServices.push(libp2pService)
  }
  logger.info(`created ${libp2pServices.length} libp2p instances`)
  return libp2pServices
}

export const timelinesInclude = (timelines: string[][], event: string): boolean => {
  if (timelines.every(timeline => timeline.includes(event))) {
    return true
  }
  return false
}
