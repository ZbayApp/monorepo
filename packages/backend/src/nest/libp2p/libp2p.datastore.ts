import { LevelDatastore } from 'datastore-level'
import { DatabaseOptions, Level } from 'level'
import { KeyQuery } from 'interface-datastore'
import { MemoryDatastore } from 'datastore-core'

import { QuietLogger } from '@quiet/logger'

import { Libp2pDatastoreOptions } from './libp2p.types'
import { createLogger } from '../common/logger'

export class Libp2pDatastore {
  private inMemory: boolean
  private datastorePath: string | undefined
  private datastore: LevelDatastore | MemoryDatastore | undefined

  private logger: QuietLogger = createLogger('libp2p:datastore')

  constructor(options: Libp2pDatastoreOptions) {
    this.inMemory = options.inMemory
    this.datastorePath = options.datastorePath
  }

  public init(): LevelDatastore | MemoryDatastore {
    this.logger.info(`Initializing Libp2pDatastore using an ${this.inMemory ? 'in-memory' : 'level'} datastore`)
    if (this.datastore != null) {
      this.logger.warn(`Libp2pDatastore already initialized, returning existing datastore instance.`)
      return this.datastore
    }

    if (this.inMemory) {
      this.datastore = new MemoryDatastore()
      return this.datastore
    }

    const datastoreInit: DatabaseOptions<string, Uint8Array> = {
      keyEncoding: 'utf8',
      valueEncoding: 'buffer',
      createIfMissing: true,
      errorIfExists: false,
      version: 1,
    }

    if (this.datastorePath == null) {
      throw new Error(`Must provide a datastorePath if using LevelDatastore`)
    }

    const datastoreLevelDb = new Level<string, Uint8Array>(this.datastorePath, datastoreInit)
    this.datastore = new LevelDatastore(datastoreLevelDb, datastoreInit)
    return this.datastore
  }

  public getDatastoreInstance(): LevelDatastore | MemoryDatastore | undefined {
    return this.datastore
  }

  public async deleteKeysByPrefix(prefix: string): Promise<string[]> {
    this.logger.info(`Deleting keys from libp2p datastore with prefix ${prefix}`)
    if (this.datastore == null) {
      this.logger.warn(`Datastore was undefined, skipping!`)
      return []
    }

    if (this.inMemory) {
      return this._deleteKeysByPrefixInMemory(prefix)
    }
    return this._deleteKeysByPrefixLevel(prefix)
  }

  private async _deleteKeysByPrefixInMemory(prefix: string): Promise<string[]> {
    const deletedKeys: string[] = []
    for await (const key of (this.datastore as MemoryDatastore)._allKeys()) {
      if (key.list()[0] === prefix) {
        this.logger.info(`Found matching key ${key.toString()}`)
        await this.datastore?.delete(key)
        deletedKeys.push(key.toString())
      }
    }

    this.logger.info(`Deleted ${deletedKeys.length} keys from in-memory datastore`)
    return deletedKeys
  }

  private async _deleteKeysByPrefixLevel(prefix: string): Promise<string[]> {
    const deletedKeys: string[] = []
    for await (const key of (this.datastore as LevelDatastore).queryKeys({
      prefix,
    } as KeyQuery)) {
      this.logger.info(`Found matching key ${key.toString()}`)
      await this.datastore?.delete(key)
      deletedKeys.push(key.toString())
    }

    this.logger.info(`Deleted ${deletedKeys.length} keys from level datastore`)
    return deletedKeys
  }
}
