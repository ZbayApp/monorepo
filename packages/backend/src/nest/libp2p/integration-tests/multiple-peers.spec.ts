import { jest } from '@jest/globals'
import { TestingModule } from '@nestjs/testing'
import { Libp2pService } from '../libp2p.service'
import { SigChainService } from '../../auth/sigchain.service'
import { Libp2pEvents } from '../libp2p.types'
import { createLogger } from '../../common/logger'
import { spawnLibp2pInstances, spawnTestModules, attachEventListeners, timelinesInclude } from './test-utils'
import { headsAreEqual, Hash } from '@localfirst/crdx'

const logger = createLogger('libp2p:multiple-peers.spec')

const N_PEERS = 3
jest.setTimeout(120_000)
describe(`Libp2pAuth with ${N_PEERS} peers`, () => {
  const teamName: string = 'team'
  const eventTimeline: string[] = []
  const eventTimelines: Array<string[]> = []
  const modules: TestingModule[] = []

  beforeAll(async () => {
    modules.push(...(await spawnTestModules(N_PEERS)))
    const sigchainServiceA = await modules[0].resolve(SigChainService)

    // Create sigChain that all other peers will join
    await sigchainServiceA.createChain(teamName, 'user0', true)
    const inviteResult = sigchainServiceA.getActiveChain().invites.createLongLivedUserInvite()

    // Initialize other chains with invite seed
    for (let i = 1; i < modules.length; i++) {
      // Create invitation from A -> B
      const sigchainService = await modules[i].resolve(SigChainService)
      await sigchainService.createChainFromInvite(`user${i}`, teamName, inviteResult.seed, true)
    }

    // Create libp2p instances
    await spawnLibp2pInstances(modules)

    // Attach event listeners to all libp2p service instances
    for (let i = 0; i < modules.length; i++) {
      eventTimelines[i] = []
      attachEventListeners(await modules[i].get(Libp2pService), eventTimeline, `${i}`)
      attachEventListeners(await modules[i].get(Libp2pService), eventTimelines[i], `${i}`)
    }
  })

  afterEach(async () => {
    // Clear event timelines
    eventTimeline.length = 0
    for (let i = 0; i < eventTimelines.length; i++) {
      eventTimelines[i].length = 0
    }
  })

  afterAll(async () => {
    // Stop all instances and close modules
    for (const module of modules) {
      const libp2pService = await module.resolve(Libp2pService)
      if (libp2pService.libp2pInstance?.status !== 'stopped') {
        await libp2pService.libp2pInstance?.stop()
      }
      await module.close()
    }
  })

  it('joins with an invitation', async () => {
    await new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('PEER_DISCONNECTED events did not occur within expected time.'))
      }, 70_000)

      const resolveIfMet = async () => {
        if (timelinesInclude(eventTimelines.slice(1), Libp2pEvents.AUTH_JOINED)) {
          for (let i = 0; i < eventTimelines.length; i++) {
            expect(eventTimelines[i]).toMatchSnapshot(`eventTimeline${i} after disconnection`)
          }
          clearTimeout(timeout)
          resolve()
        }
      }

      for (const libp2pService of modules.slice(1).map(module => module.get(Libp2pService))) {
        libp2pService.once(Libp2pEvents.AUTH_JOINED, () => {
          resolveIfMet()
        })
      }
      const libp2pService = modules[0].get(Libp2pService)
      for (let i = 1; i < modules.length; i++) {
        modules[i].get(Libp2pService).dialPeer(libp2pService.localAddress)
      }
    })
  }, 120_000)
  it('emits connected after syncing', async () => {
    logger.info('emits connected after syncing')
    await new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('PEER_DISCONNECTED events did not occur within expected time.'))
      }, 70_000)
      const resolveIfMet = async () => {
        if (timelinesInclude(eventTimelines.slice(1), Libp2pEvents.AUTH_JOINED)) {
          for (let i = 0; i < eventTimelines.length; i++) {
            expect(eventTimelines[i]).toMatchSnapshot(`eventTimeline${i} after disconnection`)
          }
          clearTimeout(timeout)
          resolve()
        }
      }
      for (const libp2pService of modules.slice(1).map(module => module.get(Libp2pService))) {
        libp2pService.once(Libp2pEvents.AUTH_CONNECTED, () => {
          resolveIfMet()
        })
      }
    })
  }, 120_000)
  it('merges graphs between all peers', async () => {
    logger.info('merges graphs between all peers')
    await new Promise<void>(resolve => {
      const resolveIfMet = async () => {
        // all peers should have the same graph head
        const heads: Hash[][] = []
        for (const module of modules) {
          const sigchainService = await module.resolve(SigChainService)
          const head = sigchainService.getActiveChain().team?.graph.head
          if (head !== undefined) {
            heads.push(head)
          }
        }
        if (heads.every(head => headsAreEqual(heads[0], head))) {
          resolve()
        }
      }
      for (const libp2pService of modules.map(module => module.get(Libp2pService))) {
        libp2pService.on(Libp2pEvents.AUTH_UPDATED, () => {
          resolveIfMet()
        })
      }
    })
  }, 120_000)
  it('gracefully disconnects', async () => {
    await new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('PEER_DISCONNECTED events did not occur within expected time.'))
      }, 70_000)
      const allDisconnected = async () => {
        if (timelinesInclude(eventTimelines, Libp2pEvents.PEER_DISCONNECTED)) {
          clearTimeout(timeout)
          resolve()
        }
      }
      for (const libp2pService of modules.map(module => module.get(Libp2pService))) {
        libp2pService.once(Libp2pEvents.AUTH_DISCONNECTED, () => {
          allDisconnected()
        })
      }
      modules[0].get(Libp2pService).hangUpPeers()
    })
    for (let i = 0; i < eventTimelines.length; i++) {
      expect(eventTimelines[i]).toMatchSnapshot(`eventTimeline${i} after disconnection`)
    }
  }, 120_000)
})
