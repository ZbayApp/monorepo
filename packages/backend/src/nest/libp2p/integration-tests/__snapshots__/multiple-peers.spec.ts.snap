// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Libp2pAuth with 3 peers intializes chains when peers connect: eventTimeline0 after disconnection 1`] = `
[
  "peerConnected",
  "peerConnected",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:awaitingIdentityChallenge,verifyingTheirIdentity:done",
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:awaitingIdentityAcceptance,verifyingTheirIdentity:done",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:awaitingIdentityChallenge,verifyingTheirIdentity:done",
]
`;

exports[`Libp2pAuth with 3 peers intializes chains when peers connect: eventTimeline1 after disconnection 1`] = `
[
  "peerConnected",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:authenticating:awaitingInvitationAcceptance",
  "authJoined",
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:done,verifyingTheirIdentity:awaitingIdentityProof",
  "authStateChanged:negotiating",
]
`;

exports[`Libp2pAuth with 3 peers intializes chains when peers connect: eventTimeline2 after disconnection 1`] = `
[
  "peerConnected",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:authenticating:awaitingInvitationAcceptance",
  "authJoined",
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:done,verifyingTheirIdentity:awaitingIdentityProof",
]
`;

exports[`Libp2pAuth with 3 peers joins with an invitation: eventTimeline0 after disconnection 1`] = `
[
  "peerConnected",
  "peerConnected",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:awaitingIdentityChallenge,verifyingTheirIdentity:done",
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:awaitingIdentityAcceptance,verifyingTheirIdentity:done",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:awaitingIdentityChallenge,verifyingTheirIdentity:done",
]
`;

exports[`Libp2pAuth with 3 peers joins with an invitation: eventTimeline1 after disconnection 1`] = `
[
  "peerConnected",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:authenticating:awaitingInvitationAcceptance",
  "authJoined",
  "authStateChanged:authenticating:checkingIdentity:provingMyIdentity:done,verifyingTheirIdentity:awaitingIdentityProof",
  "authStateChanged:negotiating",
]
`;

exports[`Libp2pAuth with 3 peers joins with an invitation: eventTimeline2 after disconnection 1`] = `
[
  "peerConnected",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:awaitingIdentityClaim",
  "authStateChanged:authenticating:awaitingInvitationAcceptance",
  "authJoined",
]
`;
